core/include/kernel_defines.h

```diff
@@ -187,6 +187,20 @@ extern "C" {
  */
 #define IS_USED(module) IS_ACTIVE(module)
 
+/**
+ * @def         KERNEL_VERSION(mayor, minor, patch, extra)
+ * @brief       Generates a 64 bit variable of a release version
+ *              Comparisons to this only apply to released branches
+ *
+ * @param[in]   a   Mayor version of the release
+ * @param[in]   b   Minor version of the release
+ * @param[in]   c   Patch level of the release
+ * @param[in]   d   Extraversion, user defined
+ *
+ * @returns     A machine readable version variable
+ */
+#define KERNEL_VERSION(a, b, c, d) (((uint64_t)(a) << 48) + ((uint64_t)(b) << 32) + \
+                                    ((uint64_t)(c) << 16) + (d))
```

Comment:
We discussed a release flag offline (which is set if RIOT_VERSION is a release, and unset, when its still in the `devel` version, so

```suggestion
#define KERNEL_VERSION(a, b, c, d, e)  (((uint64_t)(a) << 48) + ((uint64_t)(b) << 32) + \
                                        ((uint64_t)(c) << 17) + ((uint64_t)(d) << 16) + (e))
```

Alternatively both c and d could be shifted by 16. As we won't have a release each month this could be feasible.

Reply:
BTW, why `KERNEL_VERSION`? I like the name proposed in the title more `RIOT_VERSION_CODE`, as the RIOT version encompasses more than just changes to `core`.

Reply:
Is there a reason this is a macro (vs. static inline)?

Reply:
Yes so it can be used in the preprocessor

Reply:
> Is there a reason this is a macro (vs. static inline)?

There might be use cases where e.g. a newly introduced `struct` member should be used on newer versions, but cannot be used on older versions. Handling this will only work with the preprocessor to the best of my knowledge. Hence, it makes IMO sense to make this available to the preprocessor.

But I don't think that the C style casts used in the expression will actually allow us using it in the preprocessor anyway. So if the casts are going to stay, I would also prefer a C function over a macro.

Reply:
> I would also prefer a C function over a macro.

Then what would be the point of this function?

Reply:
Can someone do a quick test, if the preprocessor is fine with (or alternatively without) casts?

Reply:
They break but adding `0ULL` appears to work